cpp/include/cudf/dfa/superstate.hpp

```diff
@@ -0,0 +1,224 @@
+   */
+  inline constexpr superstate(State state) : _data(0)
+  {
+    for (auto i = 0; i < N; i++) { _data |= static_cast<Data>(state) << (i * BITS); }
+  }
+
+  /**
+   * @brief create a superstate with a specific underlying value
+   * @param data the underlying value
+   */
+  inline constexpr superstate(Data data) : _data(data) {}
+
+  inline constexpr Data data() const { return _data; }
+
+  /**
+   * @brief converts to a single state using the least significant bits of the underlying data
+   *
+   * @return State a state created from the least significant bits of the underlying data
+   */
+  inline constexpr operator State() const { return static_cast<State>(_data & MASK); }
+
+  /**
+   * @brief transitions each underlying state using the provided instruction
+   *
+   * @note the + operator is called N times, once for each state.
+   *
+   * @param rhs the instruction used to transition all states
+   * @return the result of all transitions
+   */
+  inline constexpr superstate operator+(Instruction rhs) const
+  {
+    superstate result(0);
+    for (auto i = 0; i < N; i++) { result.set<false>(i, get(i) + rhs); }
+    return result;
+  }
+
+  /**
+   * @brief concatenates all underlying states and transitions in to a new superstate
+   *
+   * @note this operator is associative (but non-commutative) and can be used in prefix sums / scans
+   *
+   * @param rhs a superstate representing a series of transitions
+   * @return constexpr superstate
+   */
+  inline constexpr superstate operator+(superstate rhs) const
+  {
+    superstate result(0);
+    for (auto i = 0; i < N; i++) { result.set<false>(i, rhs.get(get(i))); }
+    return result;
+  }
+
+  /**
+   * @brief compares the underlying data equality. all represented states must match exactly.
+   *
+   * @param rhs value to be compared
+   * @return true if all represented states match
+   * @return false if at least one represented state does not match
+   */
+  inline constexpr bool operator==(superstate rhs) const { return _data == rhs._data; }
+
+  /**
+   * @brief
+   *
+   * @param rhs value to be compared
+   * @return false if all represented states match
+   * @return true if at least one represented state does not match
+   */
+  inline constexpr bool operator!=(superstate rhs) const { return _data != rhs._data; }
+
+  /**
+   * @brief get the final state after all represented transitions have been applied to the input
+   *
+   * @param state starting state before any transitions
+   * @return State ending state after all transitions
+   */
+  inline constexpr State get(State state) const { return get(static_cast<uint8_t>(state)); }
+
+ private:
+  inline constexpr State get(uint8_t n) const
+  {
+    return static_cast<State>((_data >> n * BITS) & MASK);
+  }
+
+  template <bool reset = true>
+  inline constexpr void set(uint8_t n, State state)
+  {
+    if (reset) { _data &= ~(MASK << n * BITS); }
+    _data |= static_cast<Data>(state) << n * BITS;
+  }
```

Comment:
This feels like a Do One Thing violation and there should just be separate `set` and `reset` functions.